[SR] cover static runtime for ops that uses getitem instead of listunpack

lakshmananrm1993 opened the pull request:
Summary: cover static runtime for ops that uses getitem instead of listunpack

Differential Revision: D55924256

swolchok commented on torch/csrc/jit/runtime/static/passes.cpp:
```diff
@@ -812,18 +812,18 @@ void EliminateTrivialEquallySplit(std::shared_ptr<torch::jit::Graph>& graph) {
 
-    Node* list_unpack_node = value_out->uses()[0].user;
-    if (list_unpack_node->kind() != prim::ListUnpack) {
+    Node* user_node = value_out->uses()[0].user;
+    if (user_node->kind() != prim::ListUnpack && user_node->kind() != aten::__getitem__) {
       continue;
     }
```
this worries me. 1) what if there is more than one `__getitem__`? as written it looks like we will just leave the other ones around?
2) what if the number of `__getitem__` calls in the input program doesn't match the size of the output list?

lakshmananrm1993 replied:
i changed it to work only fb::equally_split. This has an argument num_splits and I am using that to do all the following checks

1. same as number of uses of this module
2. all the uses are get_item
3. get_item uses is not duplicated

only after this I do the fuse

swolchok commented on torch/csrc/jit/runtime/static/passes.cpp:
```diff
@@ -866,36 +866,7 @@ bool shouldNotFuseListUnpackSpecialCase(const Node* node) {
-          "fb::split_and_squeeze",
-          "static_runtime::fused_split_and_squeeze_copy")};
-
+void FuseListUnpack(std::shared_ptr<torch::jit::Graph>& graph, const c10::FastMap<c10::Symbol, c10::Symbol> unfused_to_fused) {
```
pass the map by const ref, not by value (missing `&`)

swolchok commented on torch/csrc/jit/runtime/static/passes.cpp:
```diff
@@ -950,6 +921,113 @@ void FuseListUnpack(std::shared_ptr<torch::jit::Graph>& graph) {
 
+
+// Similar to FuseListUnpack but the operators results are being accessed by get_item rather than ListUnpack
+void FuseGetItemUnpack(std::shared_ptr<torch::jit::Graph>& graph, const c10::FastMap<c10::Symbol, c10::Symbol> unfused_to_fused) {
```
ditto const ref

lakshmananrm1993 replied:
fixed

swolchok commented on torch/csrc/jit/runtime/static/passes.cpp:
```diff
@@ -950,6 +921,113 @@ void FuseListUnpack(std::shared_ptr<torch::jit::Graph>& graph) {
+
+// Similar to FuseListUnpack but the operators results are being accessed by get_item rather than ListUnpack
+void FuseGetItemUnpack(std::shared_ptr<torch::jit::Graph>& graph, const c10::FastMap<c10::Symbol, c10::Symbol> unfused_to_fused) {
+  // replacement contains (old_node, new_node, list_unpack_node)
```
It would probably be clearer to define a struct rather than use a tuple.

lakshmananrm1993 replied:
changed it to struct

swolchok commented on torch/csrc/jit/runtime/static/passes.cpp:
```diff
@@ -950,6 +921,113 @@ void FuseListUnpack(std::shared_ptr<torch::jit::Graph>& graph) {
+
+    std::vector<std::tuple<Node*, ArrayRef<Value*>>> user_to_remove;
+
+    for(auto index = 0; index < value_out->uses().size(); index++) {
```
use c10::irange

swolchok commented on torch/csrc/jit/runtime/static/passes.cpp:
```diff
@@ -950,6 +921,113 @@ void FuseListUnpack(std::shared_ptr<torch::jit::Graph>& graph) {
+      continue;
+    }
+
+    replacement.emplace_back(node, new_sym, user_to_remove);
```
missing std::move() of `user_to_remove`

lakshmananrm1993 replied:
fixed

swolchok commented on torch/csrc/jit/runtime/static/passes.cpp:
```diff
@@ -950,6 +921,113 @@ void FuseListUnpack(std::shared_ptr<torch::jit::Graph>& graph) {
+} // namespace jit
+
+void FuseUnpack(std::shared_ptr<torch::jit::Graph>& graph){
+  const c10::FastMap<c10::Symbol, c10::Symbol> unfused_to_fused = {
```
this map should probably be static const

lakshmananrm1993 replied:
fixed

swolchok reviewed (changes requested):
overall concerned about it being easier to create cases it looks like this code didn't consider using `__getitem__`

lakshmananrm1993 commented:
> overall concerned about it being easier to create cases it looks like this code didn't consider using `__getitem__`

I have changed it to cover the scenarios. I do the fuse only if the following validations succeed and I changed it to work only for equally_split

same as number of uses of this module
all the uses are get_item
get_item uses is not duplicated

swolchok commented on torch/csrc/jit/runtime/static/passes.cpp:
```diff
@@ -812,18 +813,27 @@ void EliminateTrivialEquallySplit(std::shared_ptr<torch::jit::Graph>& graph) {
+    const auto num_splits_ivalue = toIValue(value_num_splits);
+
+    // check if equally_split num of splits is 1.
+    if(!num_splits_ivalue.has_value() || !num_splits_ivalue.value().isInt() || num_splits_ivalue.value().toInt() != 1) {
```
I think using `constant_as<int>()` would be briefer here (and the other places you use `toIValue`)

above:

`const auto opt_num_splits = constant_as<int>(value_num_splits);`

and here:

`if (opt_num_splits.value_or(0) != 1) {`

swolchok replied:
why is it correct to add this check to the ListUnpack case as well? was it a bug that it wasn't there before? if so, can you split this check out of this PR into a separate PR that comes with a test case demonstrating why it was broken before? If not, aren't we regressing the ListUnpack optimization?

swolchok commented on torch/csrc/jit/runtime/static/passes.cpp:
```diff
@@ -950,6 +931,149 @@ void FuseListUnpack(std::shared_ptr<torch::jit::Graph>& graph) {
+      continue;
+    }
+
+    bool should_fuse = true;
+
+    std::vector<UserToRemove> user_to_remove_list;
+
+    std::unordered_set<std::int64_t> getItem_indices;
+
+    for(auto index = 0; index < value_out->uses().size(); index++) {
```
nit: please do not put so many blank lines; keep style consistent with the surrounding code

lakshmananrm1993 replied:
removed the not needed blank lines

swolchok commented on torch/csrc/jit/runtime/static/passes.cpp:
```diff
@@ -950,6 +931,149 @@ void FuseListUnpack(std::shared_ptr<torch::jit::Graph>& graph) {
+        break;
+      }
+
+      auto getItem_indice_out = toIValue(user->inputs()[1]);
```
use `constant_as<int>()` as recommended above

lakshmananrm1993 replied:
done

swolchok commented on torch/csrc/jit/runtime/static/passes.cpp:
```diff
@@ -950,6 +931,149 @@ void FuseListUnpack(std::shared_ptr<torch::jit::Graph>& graph) {
+
+    std::vector<UserToRemove> user_to_remove_list;
+
+    std::unordered_set<std::int64_t> getItem_indices;
```
use c10::FastSet

lakshmananrm1993 replied:
done

swolchok commented on torch/csrc/jit/runtime/static/passes.cpp:
```diff
@@ -950,6 +931,149 @@ void FuseListUnpack(std::shared_ptr<torch::jit::Graph>& graph) {
+
+      LOG(INFO) << "Found fb::equally_split node2: " << user->kind().toQualString();
+
+      getItem_indice_out.emplace(getItem_indice_out);
```
this does nothing

lakshmananrm1993 replied:
it should have been getItem_Indices. Updated it now

swolchok commented on torch/csrc/jit/runtime/static/passes.cpp:
```diff
@@ -950,6 +931,149 @@ void FuseListUnpack(std::shared_ptr<torch::jit::Graph>& graph) {
+        break;
+      }
+
+      LOG(INFO) << "Found fb::equally_split node2: " << user->kind().toQualString();
```
why node2 and not just node?

lakshmananrm1993 replied:
this was testing comment. removed it

swolchok commented on torch/csrc/jit/runtime/static/passes.cpp:
```diff
@@ -950,6 +931,149 @@ void FuseListUnpack(std::shared_ptr<torch::jit::Graph>& graph) {
+    replacement_list.emplace_back(replacement);
+  }
+
+  for(const auto& replacement : replacement_list) {
+    auto* new_node = graph->create(replacement.new_sym, 0);
+    for (Value* in : replacement.node_to_be_fused->inputs()) {
+        new_node->addInput(in);
+      }
+    for(const auto& user_to_remove_node : replacement.user_to_remove) {
+      for (Value* out : user_to_remove_node.user_outputs) {
+        Value* new_out = new_node->addOutput();
+        new_out->copyMetadata(out);
+        out->replaceAllUsesWith(new_out);
+      }
+      user_to_remove_node.user->destroy();
+    }
+    new_node->insertAfter(replacement.node_to_be_fused);
+    replacement.node_to_be_fused->destroy();
+  }
```
it would be easier for me to review this for correctness if it ended up looking the same as the FuseListUnpack path. If you need to send a separate stacked PR to get rid of the std::tuple in FuseListUnpack, go for it

lakshmananrm1993 commented on torch/csrc/jit/runtime/static/passes.cpp:
```diff
@@ -950,6 +931,149 @@ void FuseListUnpack(std::shared_ptr<torch::jit::Graph>& graph) {
+
+    const Value* value_num_splits = node->inputs()[1];
+
+    // check if equally_split num of splits is equal to the number of output uses
```
removed

lakshmananrm1993 commented on torch/csrc/jit/runtime/static/passes.cpp:
```diff
@@ -950,6 +931,149 @@ void FuseListUnpack(std::shared_ptr<torch::jit::Graph>& graph) {
+    const Value* value_num_splits = node->inputs()[1];
+
+    // check if equally_split num of splits is equal to the number of output uses
+    if(!toIValue(value_num_splits).has_value() || !toIValue(value_num_splits).value().isInt() || toIValue(value_num_splits).value().toInt() != value_out->uses().size()) {
```
replaced it with constant_as<int>()

lakshmananrm1993 commented on torch/csrc/jit/runtime/static/passes.cpp:
```diff
@@ -950,6 +931,149 @@ void FuseListUnpack(std::shared_ptr<torch::jit::Graph>& graph) {
+// Its only for fb::equally_split operator
+// the operators results are being accessed by get_item rather than ListUnpack
+void FuseEquallySplitGetItemUnpack(std::shared_ptr<torch::jit::Graph>& graph) {
+  // replacement contains (old_node, new_node, list_unpack_node)
```
removed

lakshmananrm1993 commented on torch/csrc/jit/runtime/static/passes.cpp:
```diff
@@ -812,18 +813,27 @@ void EliminateTrivialEquallySplit(std::shared_ptr<torch::jit::Graph>& graph) {
+
+    const auto num_splits_ivalue = toIValue(value_num_splits);
+
+    // check if equally_split num of splits is 1.
```
removed

swolchok reviewed (approved):
propagating internal accept to github

lakshmananrm1993 commented:
@pytorchbot merge

lakshmananrm1993 committed:
[SR] cover static runtime for ops that uses getitem instead of listunpack (#124148)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/124148

cover static runtime for ops that uses getitem instead of listunpack.

Reviewed By: snabelkabiya, swolchok

Differential Revision: D55924256
